app/cluster/state/etcd.go

```diff
@@ -0,0 +1,289 @@
+	})
+
+	return &ETCDConfig{
+		Endpoints:            endpoints,
+		Namespace:            namespace,
+		ServerIndex:          serverIndex,
+		etcdWatchTimeout:     etcdWatchTimeout,
+		dialTimeout:          dialTimeout,
+		dialKeepAliveTime:    keepaliveTime,
+		dialKeepAliveTimeout: keepaliveTimeout,
+		etcdGetTimeout:       etcdGetTimeout,
+	}
+}
+
+type ETCDManager struct {
+	Config  *ETCDConfig
+	Client  *clientv3.Client
+	once    sync.Once
+	initErr error
+}
+
+func (manager *ETCDManager) init() error {
+	manager.once.Do(func() {
+		cli, err := clientv3.New(clientv3.Config{
+			Endpoints:            manager.Config.Endpoints,
+			DialTimeout:          manager.Config.dialTimeout,
+			DialKeepAliveTime:    manager.Config.dialKeepAliveTime,
+			DialKeepAliveTimeout: manager.Config.dialKeepAliveTimeout,
+			DialOptions: []grpc.DialOption{
+				grpc.WithBlock(), // block until the underlying connection is up
+			},
+		})
+		if err != nil {
+			manager.initErr = err
+			return
+		}
+		manager.Client = cli
+	})
+
+	return manager.initErr
+}
+
+// Ping ensures the connection to etcd is alive
+func (manager *ETCDManager) Ping() error {
+	if err := manager.init(); err != nil {
+		return err
+	}
+
+	_, err := manager.Client.Cluster.MemberList(context.Background())
+	if err != nil {
+		return fmt.Errorf("ping: get cluster member list: %w", err)
+	}
+
+	return nil
+}
+
+func (manager *ETCDManager) prepareMode(raw []byte) servermode.ChangeEvent {
+	var req modeRequestValue
+	err := json.Unmarshal(raw, &req)
+	if err != nil {
+		return servermode.ChangeEventError(fmt.Errorf("unmarshal mode request: %w", err))
+	}
+
+	mode := req.Mode
+	if !mode.Valid() {
+		return servermode.ChangeEventError(fmt.Errorf("invalid mode: %s", mode))
+	}
+
+	return servermode.NewChangeEvent(
+		mode,
+		func() error {
+			ackValue, err := json.MarshalToString(modeAckValue{
+				Status: mode,
+			})
+			if err != nil {
+				return fmt.Errorf("marshal ack value: %w", err)
+			}
+			_, err = manager.Client.Put(context.Background(), req.AckKey, ackValue)
+			if err != nil {
+				return fmt.Errorf("put value to ack key %q: %w", req.AckKey, err)
+			}
+			return err
+		})
+}
+
+func errChModeRequest(err error) <-chan servermode.ChangeEvent {
+	ch := make(chan servermode.ChangeEvent, 1)
+	ch <- servermode.ChangeEventError(err)
+	close(ch)
+	return ch
+}
+
+func (manager *ETCDManager) ServerMode(ctx context.Context) <-chan servermode.ChangeEvent {
+	if err := manager.init(); err != nil {
+		return errChModeRequest(err)
+	}
+
+	modeRequestKey := fmt.Sprintf(modeRequestKeyPattern, manager.Config.Namespace, manager.Config.ServerIndex)
+
+	resultChan := make(chan servermode.ChangeEvent, 1)
+	resp, err := manager.Client.Get(ctx, modeRequestKey)
+	if err != nil {
+		return errChModeRequest(err)
+	}
+
+	if len(resp.Kvs) == 0 {
+		return errChModeRequest(errors.New("no workspace found"))
```

Comment:
This might be a plausible case when we are using manual mode, We would be first scaling up the server with empty workspaces and then would move a workspace on to that empty server

Reply:
Should we be returning an error ? I feel we can just ignore and move ahead

Reply:
This seems to be an incorrect error at server Mode... If we don't get any mode from ETCD, we should start by default in Degraded

Reply:
If this is possible to happen we need to revisit this. In my view is quite unsafe to start a watch without getting the key (revision specifically) first.
 It might be possible between get and watch, for a put to happen. This is going to be lost if I remember correctly.

Let's rethink about it

Reply:
We can start from the initial revision right? for this particular key. If i remember it correctly there is also something called mod revision at a key level